File: Week 6 - Python/sentimental-credit/credit.py
from cs50 import get_int


def main():
    number = get_int('Number: ')
    result = check(number)
    if result == 0 and visa(number) == True:
        print('VISA')
    elif result == 0 and amex(number) == True:
        print('AMEX')
    elif result == 0 and master(number) == True:
        print('MASTERCARD')
    else:
        print('INVALID')


def check(number):
    # Luhn's algorithm
    sum1 = 0
    sum2 = 0

    while number > 0:
        rest = number % 10  # last digit
        sum1 += rest  # sum of not multiplyed digits
        number = int(number / 10)  # delete last digit
        rest = number % 10  # new last digit
        rest = rest * 2
        sum2 = sum2 + int(rest / 10)  # if result >= 10
        sum2 = sum2 + (rest % 10)
        number = int(number / 10)
    result = (sum1 + sum2) % 10
    return result


def visa(number):
    # 13 digits beginning with 4
    if number >= 4000000000000 and number < 5000000000000:
        return True
    # 16 digits beginning with 4
    elif number >= 4000000000000000 and number < 5000000000000000:
        return True
    else:
        return False


def amex(number):
    # 15 digits starting with 34
    if number >= 340000000000000 and number < 350000000000000:
        return True
    # 15 digits starting with 37
    elif number >= 370000000000000 and number < 380000000000000:
        return True
    else:
        return False


def master(number):
    # 16 digits starting with 51,...,56
    if number >= 5100000000000000 and number < 5700000000000000:
        return True
    else:
        return False


main()

File: Week 1 - C/credit/credit.c
#include <cs50.h>
#include <stdio.h>
// #include <stdbool.h> enables bool instead of _bool and false and true as alias for 0 1 already included in cs50.h

int check();
int visa();
int amex();
int master();

int main(void)
{
    long number;
    int checksume;

    number = get_long_long("Number:");
    checksume = check(number);
    if (checksume % 10 == 0)
    {
        if (visa(number) == 1)
        {
            printf("VISA\n");
        }
        else if (amex(number) == 1)
        {
            printf("AMEX\n");
        }
        else if (master(number) == 1)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

}

int check(long long i)
{
    int rest;
    int sum1 = 0;
    int sum2 = 0;
    int result;
    do
    {
        rest = i % 10;              //letzte stelle auslesen
        sum1 = sum1 + rest;         //zu summe der nicht miltiplizierten zahlen rechnen
        i = i / 10;                 //letzte stelle löschen mit /10 teilen
        rest = i % 10;              //vorletzte stelle  auslesen
        rest = rest * 2;            //mit 2 multiplizieren
        sum2 = sum2 + (rest / 10);  //fals 2 stellige zahl ergebnis vorletzte stelle
        sum2 = sum2 + (rest % 10);  //letzte stelle
        i = i / 10;

    }
    while (i > 0);
    result = sum1 + sum2;
    return result;
}

int visa(long long i)
{

    if (i > 4000000000000 && i < 5000000000000) //13 stellen beginnt mit 4
    {
        return 1;
    }
    else if (i > 4000000000000000 && i < 5000000000000000) //16stellen beginnt mit 4
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

int amex(long long i)
{

    if (i > 339999999999999 && i < 350000000000000) //15 stellen lang beginnt mit 34
    {
        return 1;
    }
    else if (i > 369999999999999 && i < 380000000000000) //15 stellen beginnt mit 37
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

int master(long long i)
{

    if (i > 5099999999999999 && i < 5600000000000000) //16 stellen beginnt mit 51-56
    {
        return 1;
    }
    else
    {
        return 0;
    }
}